src/main/java/net/minecraftforge/common/MapGenStructureManager.java

```diff
@@ -0,0 +1,159 @@
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * Manage MapGenStructure for the ChunkGenerators.
+ */
+public class MapGenStructureManager
+{
+    private static final Set<String> allStructureNames = new TreeSet<>();
+    private final Map<String, MapGenStructure> structureMap;
+
+    public MapGenStructureManager(List<MapGenStructure> structures)
+    {
+        ImmutableMap.Builder<String, MapGenStructure> builder = ImmutableMap.builder();
+        for (MapGenStructure s : structures)
+        {
+            String name = s.getStructureName();
+            builder.put(name, s);
+            addStructureName(name);
+        }
+        structureMap = builder.build();
+    }
+
+    /**
+     * Add a structure name to the locate command autocomplete list.
+     * <p>
+     * Is automatically done on construction
+     *
+     * @param structure Name of the structure
+     */
+    protected static void addStructureName(String structure)
+    {
+            allStructureNames.add(structure);
+    }
+
+    /**
+     * @return A new array containing both the given strings as well as the registered structure names
+     */
+    public static String[] appendStructureNames(String[] vanilla)
+    {
+        return ArrayUtils.addAll(vanilla, allStructureNames.toArray(new String[0]));
+    }
+
+    /**
+     * Get the nearest structure with the given name to the given world position.
+     * Returns null if the structure is not found
+     */
+    @Nullable
+    public BlockPos getNearestStructurePos(World worldIn, String structureName, BlockPos position, boolean findUnexplored)
+    {
+        if (structureMap.containsKey(structureName))
+        {
+            return structureMap.get(structureName).getNearestStructurePos(worldIn, position, findUnexplored);
+        }
+        return null;
+    }
+
+    /**
+     * Check if the given world position is within a structure with the given name.
+     * Returns false if the structure is not found
+     */
+    public boolean isInsideStructure(World worldIn, String structureName, BlockPos pos)
+    {
+        if (structureMap.containsKey(structureName))
+        {
+            return structureMap.get(structureName).isInsideStructure(pos);
+        }
+        return false;
+    }
+
+    /**
+     * Get the creatures of the given type the structure at the given position would like to spawn.
+     *
+     * @param defaults List of spawn entries. Should contain the (biome's) default entries. Won't be modified
+     */
+    @Nonnull
+    public List<Biome.SpawnListEntry> getSpawns(List<Biome.SpawnListEntry> defaults, World world, EnumCreatureType creatureType, BlockPos pos)
+    {
+        for (MapGenStructure structure : structureMap.values())
+        {
+            if (structure instanceof ISpawningStructure && structure.isInsideStructure(pos))
```

Comment:
The usage of `isInsideStructure` should be `isPositionInStructure` here, the former is used to check if a position is in one of the components of a structure, currently only used by the advancement location predicate. While the latter is for determining if a position is eligible for spawning, with the default behavior being check if the position is within the surrounding bounds of a structure.

Reply:
Yes, thank you. Changed. Vanilla seems to use a different kind of check for every Vanilla spawning structure though, slightly strange.

Had to manipulated the StructureTest MapGenStructure slightly.  isPositionInStructure is always false here because we are generating an empty/dummy structure.